core/services/prom_reporter.go

```diff
@@ -0,0 +1,119 @@
+package services
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/store"
+	"github.com/smartcontractkit/chainlink/core/store/models"
+	"go.uber.org/multierr"
+	"gopkg.in/guregu/null.v4"
+)
+
+//go:generate mockery --name PrometheusBackend --output ../internal/mocks/ --case=underscore
+type (
+	promReporter struct {
+		db      *sql.DB
+		backend PrometheusBackend
+	}
+
+	PrometheusBackend interface {
+		SetUnconfirmedTransactions(int64)
+		SetMaxUnconfirmedBlocks(int64)
+	}
+
+	defaultBackend struct{}
+)
+
+var (
+	promUnconfirmedTransactions = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "unconfirmed_transactions",
+		Help: "Number of currently unconfirmed transactions",
+	})
+	promMaxUnconfirmedBlocks = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "max_unconfirmed_blocks",
+		Help: "The max number of blocks any currently unconfirmed transaction has been unconfirmed for",
+	})
+)
+
+func (defaultBackend) SetUnconfirmedTransactions(n int64) {
+	promUnconfirmedTransactions.Set(float64(n))
+}
+
+func (defaultBackend) SetMaxUnconfirmedBlocks(n int64) {
+	promMaxUnconfirmedBlocks.Set(float64(n))
+}
+
+func NewPromReporter(db *sql.DB, opts ...PrometheusBackend) store.HeadTrackable {
+	var backend PrometheusBackend
+	if len(opts) > 0 {
+		backend = opts[0]
```

Comment:
Is the idea we might have multiple backends one day but now we only support one?

Reply:
The idea is just to make the argument optional.

Reply:
I see. A minor nit in that case: it'd be a little surprising as a user of NewPromReporter if I pass in multiple PrometheusBackend's and only the first one gets used. Another way to do the optional arg where its clear only 1 is allowed is to accept a *PrometheusBackend and if its nil you use the defaultBackend.